xds/internal/test/xds_server_integration_test.go

```diff
@@ -129,35 +119,68 @@ func (s) TestServerSideXDS(t *testing.T) {
 					{ "type": "insecure" }
 				]
 			}]
-		}`, fs.Address)
+		}`, serverURI)
 	bootstrapFileName := path.Join(tmpdir, "bootstrap")
 	if err := ioutil.WriteFile(bootstrapFileName, []byte(bootstrapContents), os.ModePerm); err != nil {
 		t.Fatalf("failed to write bootstrap file: %v", err)
 	}
-	if err := os.Setenv("GRPC_XDS_BOOTSTRAP", bootstrapFileName); err != nil {
-		t.Fatalf("failed to update bootstrap environment variable: %v", err)
-	}
+
+	origBootstrapFileName := env.BootstrapFileName
+	env.BootstrapFileName = bootstrapFileName
 	t.Logf("Create bootstrap file at %s with contents\n%s", bootstrapFileName, bootstrapContents)
+	return func() { env.BootstrapFileName = origBootstrapFileName }
+}
+
+// TestServerSideXDS is an e2e tests for xDS use on the server. This does not
+// use any xDS features because we have not implemented any on the server side.
+func (s) TestServerSideXDS(t *testing.T) {
+	// Spin up a fake xDS management server on a local port.
+	// TODO(easwars): Switch to using the server from envoy-go-control-plane.
+	fs, cleanup, err := fakeserver.StartServer()
+	if err != nil {
+		t.Fatalf("failed to start fake xDS server: %v", err)
+	}
+	defer cleanup()
+	t.Logf("Started xDS management server at %s", fs.Address)
+
+	// Setup the fakeserver to respond with a Listener resource.
+	setupListenerResponse(fs.XDSResponseChan, listenerName)
+	// Create a bootstrap file in a temporary directory.
+	defer setupBootstrapFile(t, fs.Address)()
 
 	// Initialize a gRPC server which uses xDS, and register stubServer on it.
 	server := xds.NewGRPCServer()
 	testpb.RegisterTestServiceServer(server, &testService{})
-	defer server.Stop()
 
-	go server.Serve(xds.ServeOptions{Network: "tcp", Address: localAddress})
-
-	// Create a clientconn and make a successful RPC
-	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
-	defer cancel()
-	cc, err := grpc.DialContext(ctx, localAddress, grpc.WithTransportCredentials(insecure.NewCredentials()))
-	if err != nil {
-		t.Fatalf("failed to dial local test server: %v", err)
+	errCh := make(chan error, 1)
+	go func() {
+		defer server.Stop()
+
+		// Create a clientconn and make a successful RPC
+		ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+		defer cancel()
+		cc, err := grpc.DialContext(ctx, localAddress, grpc.WithTransportCredentials(insecure.NewCredentials()))
+		if err != nil {
+			errCh <- fmt.Errorf("failed to dial local test server: %v", err)
+			return
+		}
+		defer cc.Close()
+
+		client := testpb.NewTestServiceClient(cc)
+		if _, err := client.EmptyCall(ctx, &testpb.Empty{}, grpc.WaitForReady(true)); err != nil {
+			errCh <- fmt.Errorf("rpc EmptyCall() failed: %v", err)
+			return
+		}
+		errCh <- nil
+	}()
+
+	opts := xds.ServeOptions{Network: "tcp", Address: localAddress}
+	if err := server.Serve(opts); err != nil {
+		t.Fatalf("Serve(%+v) failed: %v", opts, err)
 	}
```

Comment:
Maybe make this happen in the goroutine?  Just `t.Errorf` this and remove `errCh`?

Reply:
So, I initially had the `Serve` in a separate goroutine and the other stuff in the main test goroutine. And for some reason that just wasn't working on GA. The `Serve` goroutine never got scheduled and the `Dial` never succeeded  (because the server actually starts listening on that port only after `Serve` receives a Listener response).

Reply:
This doesn't make sense to me.  `Dial` should hit many synchronization points, so the `Serve` goroutine should get scheduled even if there's only 1 core.  I think this should be understood further.  Can you reproduce this failure mode locally with `-cpu 1`?